### ISSUE TITLE
Generating bindings automatically

### ISSUE THREAD

**Issue by chances** [gecko0307/bindbc-wgpu#10]
@gecko0307 From your [comment](https://github.com/gecko0307/bindbc-wgpu/issues/7#issuecomment-721411747) in #7, the wgpu function declarations are being hand-written?

## Alternatives

https://wiki.dlang.org/Bindings#Binding_generators

- [cbindgen](https://github.com/eqrion/cbindgen#readme)
    @moon-chilled's common format issue: eqrion/cbindgen#616
- [dpp](https://github.com/atilaneves/dpp)
    I was seeing success using it with `wgpu.h`
    **Con:** It generates a [bloated D file](https://github.com/chances/wgpu-d/blob/d372879c86f63bdc305b41037a8b297149b3fddd/source/wgpu/bindings.d).
- [dstep](https://github.com/jacob-carlborg/dstep)
- ~~[htod](https://dlang.org/htod.html)~~
    Unmaintained and doesn't work with `wgpu.h`

**Comment by moon-chilled** [comment]
I think a better alternative would be hooking rust's own bindgen functionality to generate d bindings.

**Comment by chances** [comment]
Perhaps an abstraction of [cbindgen](https://github.com/eqrion/cbindgen) that transforms its Intermediate Representation into a D module with the necessary `extern`s and whatnot?

[cbindgen's Internals](https://github.com/eqrion/cbindgen/blob/master/internals.md)

**Comment by moon-chilled** [comment]
Actually, upon thinking about the matter a bit more, I think even better would be for cbindgen to dump its internal representation into some common format—json or similar—which can be consumed by any language.  (Like how bindings are automatically generated for xcb and opengl based on xml specs.)  I'll file a ticket on cbindgen later.

**Comment by chances** [comment]
I just updated the bindings manually for #12 (ugh, 😝️).

@moon-chilled Have you made any progress experimenting with cbindgen?

**Comment by cyrusmsk** [comment]
Finally available with ImportC? https://github.com/chances/wgpu-d

**Comment by chances** [comment]
@cyrusmsk Exactly, though wgpu-d is still quite verbose in its "idiomatic" wrapping.

ImportC is magic. 😄

**Comment by kassane** [comment]
D support (WiP) - w/cpp_compat
- https://github.com/mozilla/cbindgen/issues/866

I'm still tweaking functors support and some fixes to the order of opaque structure templates.
